opm/grid/cpgrid/CpGridData.cpp

```diff
@@ -2509,6 +2606,65 @@ CpGridData::refinePatch(const std::array<int,3>& cells_per_dim, const std::array
             parent_to_children_cells, child_to_parent_faces, child_to_parent_cells};
 }
 
+bool CpGridData::mark(int refCount, const cpgrid::Entity<0>& element)
+{
+    if (refCount == -1) {
+        OPM_THROW(std::logic_error, "Coarsening is not supported yet.");
+    }
+    // Check the cell to be marked for refinement has no NNC (no neighbouring connections). Throw otherwise.
+    if (hasNNCs({element.index()}) && (refCount == 1)) {
+        OPM_THROW(std::logic_error, "NNC face on a cell containing LGR is not supported yet.");
+    }
+    assert((refCount == 0) || (refCount == 1)); // Do nothing (0), Refine (1), Coarsen (-1) not supported yet.
+    if (mark_.empty()) {
+        mark_.resize(this->size(0));
+    }
+    mark_[element.index()] = refCount;
+    return (mark_[element.index()] == refCount);
+}
+
+int CpGridData::getMark(const cpgrid::Entity<0>& element) const
+{
+    if (mark_.empty()) {
+        OPM_THROW(std::logic_error, "No element has been marked.");
+    }
+    else {
+        return mark_[element.index()]; // 1 refinement, 0 doing nothing, -1 coarsening (not supported yet)
+    }
+
+}
+
+bool CpGridData::preAdapt()
+{
+    // [Indirectly] Set mightVanish flags for elements that have been marked for refinement
+    if(mark_.empty()) {
+        return false;
+    }
+    else {
+        // Auxiliary bool. It will be rewritten in case an element has been marked for refinement
+        bool refinementOrCoarsening = false;
+        for (int elemIdx = 0; elemIdx <  this-> size(0); ++elemIdx) {
+            const auto& element = Dune::cpgrid::Entity<0>(*this, elemIdx, true);
+            const auto& elemMark = getMark(element);  // 1 (to be refined), 0 (do nothing), -1 (to be coarsened - not supported yet)
+            refinementOrCoarsening = refinementOrCoarsening || (elemMark != 0);
+            if (refinementOrCoarsening) {
+                break;
+            }
+        }
+        return refinementOrCoarsening;
```

Comment:
Might this accomplish the same?
```suggestion
        for (int elemIdx = 0; elemIdx <  this-> size(0); ++elemIdx) {
            const auto& element = Dune::cpgrid::Entity<0>(*this, elemIdx, true);
            if (getMark(element) != 0)  // 1 (to be refined), 0 (do nothing), -1 (to be coarsened - not supported yet)
                return true;
        }
        return false;
```

Reply:
Indeed, I tried to commit this suggestion and an error occurred so I'm incorporating this "by hand"